File: app/src/modules/calendar/api/service/appointment-service.ts
import { Appointment } from '@/modules/calendar/api/entities/appointment';
import { CreateAppointment } from '@/modules/calendar/api/data-transfer-objects/create-appointment';
import { UpdateAppointment } from '@/modules/calendar/api/data-transfer-objects/update-appointment';
import { http } from '@/core';
import { AppointmentBlock } from '@/modules/calendar/api/entities/appointment-block';
import moment from 'moment';
import { CalendarRange } from '@/modules/app/store/calendar/calendar-range';

export class AppointmentService {
    public async get(day: Date, range: CalendarRange): Promise<Appointment[]> {
        const res = await http.get<any[]>('/appointment', {
            params: { range: 'day', date: moment(day).format('YYYY-MM-DD') }
        });

        if (res.data == null) {
            return [];
        }

        return res.data.map(d => this._map(d));
    }

    public async createAppointment(appointment: CreateAppointment): Promise<Appointment> {
        const res = await http.post<any>('/appointment', appointment);
        return this._map(res.data);
    }

    public async updateAppointment(appointment: UpdateAppointment): Promise<Appointment> {
        const res = await http.put<any>(`/appointment/${appointment.id}`, appointment);
        return this._map(res.data);
    }

    public async deleteAppointment(id: string): Promise<void> {
        await http.delete(`/appointment/${id}`);
    }

    _map(data: any): Appointment {
        const appointment = new Appointment(data.serviceId, data.clientId, data.price, [], data.notes);

        const blocks = data.blocks.map((t: any) => {
            const block = new AppointmentBlock(t.start, t.end);
            block.appointment = appointment;
            block.id = t.id;

            return block;
        });

        appointment.blocks = blocks;
        appointment.id = data.id;

        return appointment;
    }
}

File: api/src/Domain/Calendar/Services/AppointmentService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace DetailingArsenal.Domain.Calendar {
    public interface IAppointmentService : IService {
        Task<Appointment> GetById(Guid id);
        Task<List<Appointment>> GetWithinRange(DateTime date, AppointmentRange range, User user);
        Task<Appointment> Create(CreateAppointment create, User user);
        Task Update(Appointment appointment, UpdateAppointment update);
        Task Delete(Appointment appointment);
    }

    public class AppointmentService : IAppointmentService {
        IAppointmentRepo repo;

        public AppointmentService(IAppointmentRepo repo) {
            this.repo = repo;
        }

        public async Task<Appointment> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<List<Appointment>> GetWithinRange(DateTime date, AppointmentRange range, User user) {
            if (range == AppointmentRange.Day) {
                return await repo.FindForDay(date, user);
            } else {
                return await repo.FindForWeek(date, user);
            }
        }

        public async Task<Appointment> Create(CreateAppointment create, User user) {
            var appointment = Appointment.Create(
                user.Id,
                create.ServiceId,
                create.ClientId,
                create.Price,
                create.Notes
            );

            appointment.Blocks = create.Blocks.Select(b => AppointmentBlock.Create(
                appointment.Id,
                b.Start,
                b.End
            )).ToList();

            await repo.Add(appointment);
            return appointment;
        }

        public async Task Update(Appointment appointment, UpdateAppointment update) {
            appointment.ServiceId = update.ServiceId;
            appointment.ClientId = update.ClientId;
            appointment.Price = update.Price;
            appointment.Notes = update.Notes;

            appointment.Blocks = update.Blocks.Select(b => AppointmentBlock.Create(
                appointment.Id,
                b.Start,
                b.End
            )).ToList();

            await repo.Update(appointment);
        }

        public async Task Delete(Appointment appointment) {
            await repo.Delete(appointment);
        }
    }
}